.tekton/retag-image-task.yaml

```diff
@@ -0,0 +1,108 @@
+    description: Digest of the output image; should be the same as of the input one
+  - name: IMAGE_URL
+    description: Image repository and tag of the output image
+  - name: IMAGE_REF
+    description: Image reference of the output image containing both the repository and the digest
+  steps:
+  - name: copy-image
+    image: quay.io/konflux-ci/release-service-utils:latest@sha256:48b31c88616f47328da613092e62b54b30cfbc7b3eb972d213db7b696f8edbf7
+    script: |
+      #!/usr/bin/env bash
+      set -euo pipefail
+      
+      SKOPEO_RETRIES=5
+      
+      function copy_image() {
+        local -r input_url="${1}"
+        local -r output_url="${2}"
+
+        if [[ "${input_url}" == "${output_url}" ]]; then
+          >&2 echo "Will not copy image ${input_url} into itself."
+          exit 5
+        fi
+
+        echo "Copying image from ${input_url} ..."
+        echo "... to ${output_url} ..."
+
+        # cosign copies not just cosign artifacts but also images and it understands and copies index images.
+        # --force argument is needed to prevent the command failing when it runs into (partial) previous copies with
+        # errors like the following.
+        #   Error: image "quay.io/rhacs-eng/collector:sha256-4b509fdf27761150a5a5ce519f087bc6d69af9d886a63f67f70b1e293643ea19" already exists. Use `-f` to overwrite
+        cosign copy --force "${input_url}" "${output_url}"
+
+        oras cp --recursive "${input_url}" "${output_url}"
+      }
+
+      function get_digest() {
+        local -r image_url="${1}"
+        skopeo inspect \
+          --retry-times="${SKOPEO_RETRIES}" \
+          --format '{{.Digest}}' \
+          --no-tags \
+          "docker://${image_url}"
+      }
+
+      INPUT_URL="$(params.INPUT_IMAGE_REPO):$(params.INPUT_IMAGE_TAG)"
+      OUTPUT_URL="$(params.OUTPUT_IMAGE_REPO):$(params.OUTPUT_IMAGE_TAG)"
+      
+      # This copies the images themselves.
+      # It also handles the case when the image is index on top of arch-specific containers in which case arch-specific
+      # containers are also copied (but not tags of arch-specific containers if they were additionally tagged).
+      copy_image "${INPUT_URL}" "${OUTPUT_URL}"
+
+      echo "Figuring the input image digest ..."
+      INPUT_DIGEST="$(get_digest "${INPUT_URL}")"
+
+      echo "Figuring the output image digest ..."
+      OUTPUT_DIGEST="$(get_digest "${OUTPUT_URL}")"
+
+      echo "Checking if the input image digest ${INPUT_DIGEST} is the same as the output image digest ${OUTPUT_DIGEST} ..."
+      [[ "${INPUT_DIGEST}" == "${OUTPUT_DIGEST}" ]]
+
+      # This handles the image with source code, not to be confused with the input image itself.
+      echo "Copying the image with sources ..."
+      # Tag determination borrowed from
+      # https://github.com/konflux-ci/release-service-catalog/blob/92dabb8a31c42669ab9b6ccce831487727514059/tasks/push-snapshot/push-snapshot.yaml#L158
+      SOURCES_IMAGE_TAG="${INPUT_DIGEST/:/-}.src"
```

Comment:
While this may seem fragile in spite of Konflux colleagues suddenly changing the tag format, I think this is unlikely to happen because the similar tagging scheme `sha256-<digest>.<suffix>` is used by cosign for its side-artifacts and `sha256-<digest>.src` fits it well. In case it will change, we would notice retagging failures for collector-slim and will adjust. Unfortunately, I don't know how to better protect from this.

![image](https://github.com/user-attachments/assets/8d8dd940-0c8b-48b1-ad6e-9345ebb101a4)